File: src/org/simbrain/util/piccolo/TileMap.java
package org.simbrain.util.piccolo;

import com.thoughtworks.xstream.annotations.XStreamAlias;
import com.thoughtworks.xstream.annotations.XStreamAsAttribute;
import com.thoughtworks.xstream.annotations.XStreamImplicit;
import org.piccolo2d.nodes.PImage;
import org.simbrain.world.odorworld.resources.OdorWorldResourceManager;

import java.awt.*;
import java.awt.geom.Point2D;
import java.io.File;
import java.util.ArrayList;
import java.util.List;

/**
 * Java representation of a .tmx tilemap produced by the Tiled app
 * (https://doc.mapeditor.org/en/stable/)
 *
 * A tilemap contains a list of {@link TileMapLayer} objects and a {@link TileSet}
 * object. Each layer is basically a grid of tiles, each of which points to a
 * member of a tileset, which is like a sprite sheet. To get a sense of
 * this see a sample tmx file like <code>sample.tmx</code>.
 *
 * The map returns a list of PImages, one per layer, which can be rendered in a
 * Piccolo canvas.
 *
 */
@XStreamAlias("map")
public class TileMap {

    /**
     * The TMX format version. Was “1.0” so far, and will be incremented to match minor Tiled releases.
     */
    @XStreamAsAttribute
    private String version;

    /**
     * The Tiled version used to save the file (since Tiled 1.0.1). May be a date (for snapshot builds).
     */
    @XStreamAsAttribute
    private String tiledversion;

    /**
     * The map width in tiles.
     */
    @XStreamAsAttribute
    private int width;

    /**
     * The map height in tiles.
     */
    @XStreamAsAttribute
    private int height;

    /**
     * The width of a tile.
     */
    @XStreamAsAttribute
    private int tilewidth;

    /**
     * The height of a tile.
     */
    @XStreamAsAttribute
    private int tileheight;

    /**
     * The tile set this map uses
     */
    private TileSet tileset;

    /**
     * The layers of this map.
     */
    @XStreamImplicit
    public ArrayList<TileMapLayer> layers = new ArrayList<>();

    /**
     * Layers of the rendered map images
     */
    private transient ArrayList<PImage> renderedLayers = null;

    /**
     * The background color of the map. (optional, may include alpha value since 0.15 in the form #AARRGGBB)
     * (Not used for now)
     */
    private Color backgroundcolor;

    /**
     * Create a tilemap by parsing a tmx file, which is an xml representation
     * of a tilemap.
     *
     * @param file file to parse
     * @return the tilemap object from the given file
     */
    public static TileMap create(File file) {
        return (TileMap) TMXUtils.getXStream().fromXML(file);
    }

    /**
     * Create a tilemap by parsing a tmx file, which is an xml representation
     * of a tilemap.
     *
     * @param filename name of the file to parse
     * @return the tilemap object from the given file
     */
    public static TileMap create(String filename) {
        return (TileMap) TMXUtils.getXStream().fromXML(OdorWorldResourceManager.getFileURL("tilemap/" + filename));
    }


    /**
     * Get a list of images of each layer of this map.
     * @return the list of layer images
     */
    public ArrayList<PImage> createImageList() {
        if (renderedLayers == null) {
            renderedLayers = new ArrayList<>();
            if (tileset == null) {
                tileset = new TileSet();
            }
            for (TileMapLayer l : layers) {
                renderedLayers.add(l.renderImage(tileset));
            }
        }
        return renderedLayers;
    }

    /**
     * Check if a tile with a given id exists at a specified location in
     * tile coordinates.
     *
     * @param id the id of the tile to check
     * @param x the x location in tile coordinate
     * @param y the y location in tile coordinate
     * @return true if the given tile exists in the tile stack
     */
    public boolean hasTileIdAt(int id, int x, int y) {

        // if id is less than first gid, the id could be a empty tile, so those should not be check.
        if (id < tileset.getFirstgid()) {
            return false;
        }

        return getTileStackAt(x, y).stream().anyMatch(t -> t.getId() == id);
    }

    /**
     * Check if a tile with a given id exists at a specified location in
     * pixel coordinates.
     *
     * @param id the id of the tile to check
     * @param x the x pixel location
     * @param y the y pixel location
     * @return true if the given tile exists in the tile stack
     */
    public boolean hasTileIdAtPixel(int id, double x, double y) {
        return getTileStackAtPixel(x, y).stream().anyMatch(t -> t.getId() == id);
    }

    /**
     * Check if a tile with a given id exists at a specified location in
     * pixel coordinates.
     *
     * @param id the id of the tile to check
     * @param p the location of the pixel to check
     * @return true if the given tile exists in the tile stack
     */
    public boolean hasTileIdAtPixel(int id, Point p) {
        return getTileStackAtPixel(p.getX(), p.getY()).stream().anyMatch(t -> t.getId() == id);
    }

    /**
     * Returns the "stack" of tiles at a given location as a list.
     *
     * @param x tile coordinate x
     * @param y tile coordinate y
     * @return a list of tiles at that location in the same order as in the xml file
     */
    public List<Tile> getTileStackAt(int x, int y) {
        ArrayList<Tile> stack = new ArrayList<>();

        // return empty if out of bound
        if (x < 0 || x > width || y < 0 || y > height) {
            return stack;
        }

        for (TileMapLayer l : layers) {
            stack.add(tileset.getTile(l.getTileIdAt(x, y)));
        }

        return stack;
    }

    /**
     * Returns the "stack" of tiles at a given location as a list.
     *
     * @param x pixel x location
     * @param y pixel y location
     * @return a list of tiles at that location in the same order as in the xml file
     */
    public List<Tile> getTileStackAtPixel(double x, double y) {
        Point tileCoordinate = pixelToTileCoordinate(x, y);
        int tileCoordinateX = (int) tileCoordinate.getX();
        int tileCoordinateY = (int) tileCoordinate.getY();
        return getTileStackAt(tileCoordinateX, tileCoordinateY);
    }

    /**
     * Returns the "stack" of tiles at a given location as a list.
     *
     * @param p pixel location
     * @return a list of tiles at that location in the same order as in the xml file
     */
    public List<Tile> getTileStackAtPixel(Point2D p) {
        // TODO: This returns some null tiles, not sure why
        Point tileCoordinate = pixelToTileCoordinate(p);
        int tileCoordinateX = (int) tileCoordinate.getX();
        int tileCoordinateY = (int) tileCoordinate.getY();
        return getTileStackAt(tileCoordinateX, tileCoordinateY);
    }

    /**
     * Converts pixel location to tile coordinate.
     *
     * @param x pixel x location
     * @param y pixel y location
     * @return the corresponding tile location
     */
    public Point pixelToTileCoordinate(double x, double y) {
        return new Point((int) (x / tilewidth), (int) (y / tileheight));
    }

    /**
     * Converts pixel location to tile coordinate.
     *
     * @param p pixel location
     * @return the corresponding tile location
     */
    public Point pixelToTileCoordinate(Point2D p) {
        return pixelToTileCoordinate(p.getX(), p.getY());
    }

    public int getMapHeight() {
        return height * tileheight;
    }

    public int getMapWidth() {
        return width * tilewidth;
    }

    public int getMapHeightInTiles() {
        return height;
    }

    public int getMapWidthInTiles() {
        return width;
    }

    public int getTilewidth() {
        return tilewidth;
    }

    public int getTileheight() {
        return tileheight;
    }
}

File: src/main/kotlin/org/simbrain/util/piccolo/TileMap.kt
package org.simbrain.util.piccolo

import com.thoughtworks.xstream.annotations.XStreamAlias
import com.thoughtworks.xstream.annotations.XStreamAsAttribute
import com.thoughtworks.xstream.annotations.XStreamImplicit
import org.piccolo2d.nodes.PImage
import org.simbrain.world.odorworld.events.TileMapEvents
import java.awt.Color
import java.awt.geom.Point2D
import java.awt.geom.Rectangle2D

/**
 * Java representation of a .tmx tilemap produced by the Tiled app
 * (https://doc.mapeditor.org/en/stable/)
 *
 * A tilemap contains a list of [TileMapLayer]s and a list of [TileSet]s.
 *
 * Each layer is basically a grid of tiles, each of which points to a
 * member of a tileset.
 *
 * To get a sense of this see a sample tmx file like `aris_world.tmx`.
 *
 * This object does not return a list of separate PImages, but rather one big image for each layer. See [#createImageList]
 * This image is rendered only when the map changes.
 *
 * The XStream annotations in this class allow XStream to read in a .tmx file in its default format.
 *
 */
@XStreamAlias("map")
class TileMap(width: Int, height: Int) {

    /**
     * The TMX format version. Was “1.0” so far, and will be incremented to match minor Tiled releases.
     */
    @XStreamAsAttribute
    private val version: String = "1.0"

    /**
     * The Tiled version used to save the file (since Tiled 1.0.1). May be a date (for snapshot builds).
     */
    @XStreamAsAttribute
    private val tiledversion: String = "1.0.1"

    /**
     * The map width in tiles.
     */
    @XStreamAsAttribute
    var width = width
        private set

    /**
     * The map height in tiles.
     */
    @XStreamAsAttribute
    var height = height
        private set

    /**
     * The width of a tile.
     */
    @XStreamAlias("tilewidth")
    @XStreamAsAttribute
    val tileWidth = 32

    /**
     * The height of a tile.
     */
    @XStreamAlias("tileheight")
    @XStreamAsAttribute
    val tileHeight = 32

    /**
     * Get the map height in pixels.
     */
    val mapHeight: Int
        get() = height * tileHeight

    /**
     * Get the map width in pixels.
     */
    val mapWidth: Int
        get() = width * tileWidth

    /**
     * The tile set this map uses
     */
    @XStreamImplicit
    @XStreamAlias("tileset")
    val tileSets: List<TileSet> = listOf(createDefaultTileSet())

    /**
     * Cache id ranges for each TileSet in [tileSets]
     */
    @Transient
    private var tileSetRanges = tileSets.map { it.firstgid..(it.tilecount + it.firstgid) to it }

    /**
     * Cache [Tile] objects by their ids
     */
    @Transient
    private var idTileMapping = HashMap<Int, Tile>()

    @XStreamImplicit
    @XStreamAlias("layers")
    private val _layers = mutableListOf(TileMapLayer("Default Layer", width, height, true))
    val layers: List<TileMapLayer>
        get() = _layers

    /**
     * The background color of the map. (optional, may include alpha value since 0.15 in the form #AARRGGBB)
     * (Not used for now)
     */
    private val backgroundcolor: Color? = null

    private var guiEnabled = false

    @Transient
    var events = TileMapEvents()
        private set

    /**
     * Get a list of images of each layer of this map.
     * @return the list of layer images
     */
    fun createImageList(): List<PImage> {
        guiEnabled = true
        return layers.map { it.renderImage(tileSets) }
    }

    fun renderLayer(layer: TileMapLayer) {
        layer.renderImage(tileSets)
    }

    /**
     * Edit tiles in first layer. Use with care; assumes just one layer.
     */
    @JvmOverloads
    fun setTile(x: Int, y: Int, tileID: Int, layer: TileMapLayer = layers.first()) {
        layer.setTile(x, y, tileID)
    }

    /**
     * @param suppressMissingNames by default this function prints out a warning if a tile could not be found in the
     *      tile sets. To disable that warning for specific tiles, provide their names in this list
     */
    fun setTile(x: Int, y: Int, tileStringID: String, layer: TileMapLayer = layers.first(), suppressMissingNames: List<String> = listOf()) {
        val tileID = tileSets.asSequence().map { it[tileStringID] }.first()?.id
        tileID?.let { layer.setTile(x, y, it + 1) }
            ?: if (tileStringID !in suppressMissingNames) println("warn: no tile string id [$tileStringID] found") else Unit
    }

    /**
     * Edit tile in a named layer
     */
    fun setTile(layerName: String, x: Int, y: Int, tileID: Int) {
        getLayer(layerName).setTile(x, y, tileID)
    }

    /**
     * Clear the indicated layer.
     */
    fun clear(layer: TileMapLayer = layers.first()) {
        fill(0, layer)
    }

    /**
     * Set all tiles on first layer to specified tile id.
     */
    fun fill(tileId: Int, layer: TileMapLayer = layers.first()) {
        (0 until width).forEach { i ->
            (0 until height).forEach { j ->
                layer[i, j] = tileId
            }
        }
        layer.render()
    }

    fun fill(tileLabel: String) {
        fill(tileSets.getGidFromLabel(tileLabel))
    }

    // TODO: should not be able to edit tile map layers that don't belong to this map
    fun TileMapLayer.setTile(x: Int, y: Int, tileID: Int) {
        this[x, y] = tileID
        render()
    }

    private fun TileMapLayer.render() {
        if (guiEnabled) {
            val oldRenderedImage = layerImage
            val newRenderedImage = renderImage(tileSets, true)
            events.layerImageChanged.fireAndForget(oldRenderedImage, newRenderedImage)
        }
    }

    fun addLayer(layer: TileMapLayer): TileMapLayer {
        _layers.add(layer)
        events.layerAdded.fireAndForget()
        return layer
    }

    /**
     * Check if a tile with a given id exists at a specified location in
     * pixel coordinates.
     *
     * @param id the id of the tile to check
     * @param x the x pixel location
     * @param y the y pixel location
     * @return true if the given tile exists in the tile stack
     */
    fun hasTileIdAtPixel(id: Int, x: Double, y: Double) =
            getTileStackAtPixel(x, y).any { t: Tile -> t.id == id }

    fun getTile(gid: Int) = idTileMapping.getOrPut(gid) {
        tileSetRanges
            .firstOrNull { (range, _) -> gid in range }
            ?.let { (_, tileSet) -> tileSet[gid] } ?: zeroTile
    }

    fun tileImage(gid: Int) = tileSetRanges
            .firstOrNull { (range, _) -> gid in range }
            ?.let { (_, tileSet) -> tileSet.getTileImage(gid) } ?: transparentTexture(tileWidth, tileHeight)

    /**
     * Returns the "stack" of tiles at a given location as a list.
     *
     * @param x tile coordinate x
     * @param y tile coordinate y
     * @return a list of tiles at that location in the same order as in the xml file
     */
    fun getTileStackAt(x: Int, y: Int) = layers.map { getTile(it[x, y]) }

    fun getTileStackAt(gridCoordinate: GridCoordinate) =
        getTileStackAt(gridCoordinate.x.toInt(), gridCoordinate.y.toInt())

    /**
     * Check if a given tile location contains any tiles or layers that with the collision property.
     *
     * @param x x in tile coordinate
     * @param y y in tile coordinate
     * @return true if the given location has a collision tile
     */
    fun collidingAt(x: Int, y: Int) =
            layers.filter { it.collision }.map { getTile(it[x, y]) }.any { it.id != 0 } ||
                    layers.map{ getTile(it[x,y]) }.any{ it.collision }

    /**
     * Get a Rectangle2D region of a given tile
     *
     * @param x x in tile coordinate
     * @param y y in tile coordinate
     * @return a Rectangle2D region representing the area of the tile
     */
    fun getTileBound(x: Int, y: Int) = Rectangle2D.Double(
            (x * tileWidth).toDouble(),
            (y * tileHeight).toDouble(),
            tileWidth.toDouble(),
            tileHeight.toDouble()
    )

    /**
     * Returns the "stack" of tiles at a given location as a list.
     *
     * @param x pixel x location
     * @param y pixel y location
     * @return a list of tiles at that location in the same order as in the xml file
     */
    fun getTileStackAtPixel(x: Double, y: Double) = pixelToGridCoordinate(x, y).let { (tileX, tileY) ->
        getTileStackAt(tileX, tileY)
    }

    /**
     * Returns the "stack" of tiles at a given location as a list.
     *
     * @param p pixel location
     * @return a list of tiles at that location in the same order as in the xml file
     */
    fun getTileStackAtPixel(p: Point2D) = getTileStackAtPixel(p.x, p.y)

    fun pixelToGridCoordinate(x: Double, y: Double) = GridCoordinate((x / tileWidth).toInt(), (y / tileHeight).toInt
        ()).int

    fun pixelToGridCoordinate(p: Point2D) = pixelToGridCoordinate(p.x, p.y)

    /**
     * Get layer by name.
     */
    fun getLayer(name: String) =
            layers.firstOrNull { name == it.name } ?: throw IllegalArgumentException("$name not a tile layer name")

    /**
     * Clear and resize the map to the specified size. Maintains current tilemap.
     *
     * TODO: consider creating empty map instead of using this to avoid mutability
     *
     * @param width the new width of the map
     * @param height the new height of the map
     */
    fun updateMapSize(width: Int, height: Int) {
        if (width < 0 || height < 0) return

        this.width = width
        this.height = height
        layers.forEach {
            it.clear(width, height)
            if (guiEnabled) {
                it.renderImage(tileSets, true)
            }
        }
        events.mapSizeChanged.fireAndBlock()
    }

    /**
     * See {@link org.simbrain.workspace.serialization.WorkspaceComponentDeserializer}
     */
    private fun readResolve(): Any {
        events = TileMapEvents()
        tileSetRanges = tileSets.map { it.firstgid..(it.tilecount + it.firstgid) to it }
        idTileMapping = HashMap()
        return this
    }

}

